### PR TITLE
Deduplicate muxer for a given same stream.

### PR DESCRIPTION
**Pull request by Nuhvi**
Not sure if this causes any problems I can't think of, But it sure fixes this [problem](https://gist.github.com/Nazeh/a3f1d24b597913303afcb6c568f4b042) as well as the new unit test (comment out line 248 to see it fail)

This would make this [PR](https://github.com/hypercore-protocol/corestore/pull/27) unnecessary too.

### CODE REVIEW

**Comment by mafintosh** [comment]
You should not return from a constructor, creates weird edge-cases. The `Protomux.from` helper exists for exactly this

**Comment by Nuhvi** [comment]
> You should not return from a constructor, creates weird edge-cases. The `Protomux.from` helper exists for exactly this

Fair enough, I was trying to cover cases where someone constructs Protomux with new, like [Hypercore.createProtocolStream](https://github.com/hypercore-protocol/hypercore/blob/master/index.js#L151) and obsolete the [PR in corestore ](https://github.com/hypercore-protocol/corestore/pull/27/files)

Fixed now.

**Comment by mafintosh** [comment]
@Nazeh we should change that to be Protomux.from, the one Hypercore one

File: index.js [hunk 1]
```diff
@@ -245,6 +245,8 @@ class Channel {
 
 module.exports = class Protomux {
   constructor (stream, { alloc } = {}) {
+    stream.protomux = this
```
Comment (mafintosh) [inline index.js]: if (stream.userData === null) stream.userData = this

File: index.js [hunk 2]
```diff
@@ -271,6 +273,7 @@ module.exports = class Protomux {
   }
 
   static from (stream, opts) {
+    if (stream.protomux) return stream.protomux
```
Comment (mafintosh) [inline index.js]: if (stream.userData && stream.userData.isProtomux) return stream.userData

**Comment by mafintosh** [comment]
Cool solution. In general we define the userData field so a stream can opt-in to modules doing this, vs force injecting it. Added a comment about that.

**Pull request merged**